ApplicationController.java

```
@@ -0,0 +1,470 @@
+            else if (msg.hasNoMeasures())
+                errors.add(new ValidationError("measures", "the list of measures is either missing or empty"));
+            else
+                validateMeasures(msg.getMeasures(), errors);
+        }
+
+        if (!errors.isEmpty())
+            throw new CloudValidationException(errors);
+        return msg;
+    }
+
+    private void validateMeasures(IoTMeasure[] measures, ArrayList<ValidationError> errors)
+    {
+        for (IoTMeasure measure : measures) {
+            String signerAlternateId = measure.getSignerAlternateId();
+
+            verifyToolId(signerAlternateId);
+            String key = "measures(" + signerAlternateId + ").jws";
+            if (measure.getJws() == null)
+                errors.add(new ValidationError(key, THE_FIELD_IS_MISSING_IN_PAYLOAD));
+            else if (!Pattern.matches(patternJws, measure.getJws()))
+                errors.add(new ValidationError(key, "Incorrect format of jws payload"));
+        }
+    }
+
+    private String decodeMeasure(IoTMeasure measure, DigitalIdentityModel record)
+    {
+        String errMsgPrefix = "measure[" + measure.getSignerAlternateId() + "] - ";
+
+        try {
+            JWSObject jws = JWSObject.parse(measure.getJws());
+            logger.debug("jws extracted");
+
+            // Extract device public key from the Digital Identity Signing Request
+            ECPublicKey publicKey = KeyRoutines.extractPublicKey(record.getPublicKey());
+            logger.debug("public key extracted");
+
+            // Verify the signature using the device public key
+            JWSVerifier verifier = new ECDSAVerifier(publicKey);
+            if (!jws.verify(verifier))
+                throw new CloudBasicException(ErrorCode.BAD_REQUEST, errMsgPrefix + "Java Web Signature validation failure");
+            logger.debug("jws verified");
+
+            return jws.getPayload().toString();
+        } catch(ParseException parseexc) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 414, 417.